### PR TITLE
Amend timeoutPacket to allow timeouts on any channel state.

### PR DESCRIPTION
**Pull request by DimitrisJim**
## Description

as discussed in the eng. call, this drops the check on channel state during a timeout. 

closes: #4080

### Commit Message / Changelog Entry

```bash
chore: amend timeoutPacket to allow timeouts on any channel state.
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/timeout.go
Comment (damiannolan) [inline modules/core/04-channel/keeper/timeout.go:87]: couldn't we also just extend this to allow timeouts on `OPEN` and `FLUSHING` channels?

Comment (DimitrisJim) [reply modules/core/04-channel/keeper/timeout.go:87]: that was my initial thought pre-discussion. This isn't spec'ed out in any way (i.e the check for `State != types.OPEN` also isn't there), so when we talked about it I remember us coming to the conclusion that it could be removed. Happy to wait for a review from @colin-axner when he gets back

Comment (colin-axner) [reply modules/core/04-channel/keeper/timeout.go:87]: Why would you want to restrict when packets can be timed out?

**Review by colin-axner — APPROVED** [review]

File: modules/core/04-channel/keeper/timeout.go
Comment (colin-axner) [reply modules/core/04-channel/keeper/timeout.go:87]: In theory that makes sense, but in practice it is problematic. There's a potential state where chainA has a closed channel and chainB doesn't. Until someone calls ChanCloseConfirm, all timeouts are valid, but with a restriction, after calling close confirm, timeouts are no longer valid, but that's a very loose assumption and I don't think apps are built to handle channel closes like that

My preference would be to open an issue on the spec if there's a concern. I don't see why ibc-go should be opinionated in this matter

Comment (colin-axner) [reply modules/core/04-channel/keeper/timeout.go:87]: There's also no check on TimeoutOnClose, so this restriction can be bypassed unless you added it there as well

Comment (damiannolan) [reply modules/core/04-channel/keeper/timeout.go:87]: Sounds fair enough to me!

**Review by damiannolan — APPROVED** [review]

File: modules/core/04-channel/keeper/timeout.go
Comment (colin-axner) [reply modules/core/04-channel/keeper/timeout.go:87]: I think the other point I'd make is that IBC only specifies optimistic sends as unsafe (potentially not recoverable). I think it's important to build the assumption that all sends can be reverted regardless of what occurs

**Pull request merged**
